modules/rank-eval/src/main/java/org/elasticsearch/index/rankeval/RatedRequest.java

```diff
@@ -48,32 +49,64 @@
  * */
 @SuppressWarnings("unchecked")
 public class RatedRequest extends ToXContentToBytes implements Writeable {
-    private String specId;
-    private SearchSourceBuilder testRequest;
-    private List<String> indices = new ArrayList<>();
-    private List<String> types = new ArrayList<>();
+    private String id;
+    private List<String> indices;
+    private List<String> types;
     private List<String> summaryFields = new ArrayList<>();
     /** Collection of rated queries for this query QA specification.*/
     private List<RatedDocument> ratedDocs = new ArrayList<>();
+    /** Search request to execute for this rated request, can be null, if template and corresponding params are supplied. */
+    @Nullable
+    private SearchSourceBuilder testRequest;
     /** Map of parameters to use for filling a query template, can be used instead of providing testRequest. */
+    @Nullable
     private Map<String, Object> params = new HashMap<>();
 
-    public RatedRequest() {
-        // ctor that doesn't require all args to be present immediatly is easier to use with ObjectParser
-        // TODO decide if we can require only id as mandatory, set default values for the rest?
-    }
+    public RatedRequest(String id, List<String> summaryFields, List<RatedDocument> ratedDocs,
+            SearchSourceBuilder testRequest, Map<String, Object> params) {
+        if (params != null && (params.size() > 0 && testRequest != null)) {
+            throw new IllegalArgumentException(
+                    "Ambiguous rated request: Set both, verbatim test request and test request template parameters.");
+        }
+        // No documents with same _index/_type/id allowed.
+        Set<DocumentKey> docKeys = new HashSet<>();
+        for (RatedDocument doc : ratedDocs) {
+            if (docKeys.add(doc.getKey()) == false) {
+                String docKeyToString = doc.getKey().toString().replaceAll("\n", "").replaceAll("  ", " ");
+                throw new IllegalArgumentException(
+                        "Found duplicate rated document key [" + docKeyToString + "]");
+            }
+        }
 
-    public RatedRequest(String specId, SearchSourceBuilder testRequest, List<String> indices, List<String> types,
-            List<RatedDocument> ratedDocs) {
-        this.specId = specId;
+        this.id = id;
         this.testRequest = testRequest;
-        this.indices = indices;
-        this.types = types;
-        setRatedDocs(ratedDocs);
+        if (summaryFields != null) {
+            this.summaryFields = summaryFields;
+        }
+        this.ratedDocs = ratedDocs;
+        if (params != null) {
+            this.params = params;
+        }
+    }
+    
+    public RatedRequest(String id, List<String> summaryFields, List<RatedDocument> ratedDocs, Map<String, Object> params) {
```

Comment:
Can we remove summaryFields from the ctors and make it an optional setter? That way we save a few of these constructors.